pkgs/build-support/trivial-builders.nix

```diff
@@ -318,23 +321,46 @@ rec {
   symlinkJoin =
     args_@{ name
          , paths
+         , outputs ? [ "out" ]
          , preferLocalBuild ? true
          , allowSubstitutes ? false
          , postBuild ? ""
          , ...
          }:
     let
+      secondaryOutputs = builtins.filter (o: o != "out") outputs;
+      outputToPathsAttr = output: {
+        name = output + "Paths";
+        value = lib.flatten (map (p: p.${output} or []) paths);
+      };
       args = removeAttrs args_ [ "name" "postBuild" ]
         // {
           inherit preferLocalBuild allowSubstitutes;
-          passAsFile = [ "paths" ];
-        }; # pass the defaults
+          outputs = [ "out" ] ++ secondaryOutputs;
+          passAsFile = [ "paths" ] ++ map (o: o + "Paths") secondaryOutputs;
+        } // builtins.listToAttrs (map outputToPathsAttr secondaryOutputs);
     in runCommand name args
       ''
         mkdir -p $out
         for i in $(cat $pathsPath); do
           ${lndir}/bin/lndir -silent $i $out
         done
+
+        for output in ${toString secondaryOutputs}; do
+          eval output_path=\$$output
+          mkdir -p $output_path
+          for i in $(eval cat \$${output}PathsPath); do
+            ${lndir}/bin/lndir -silent $i $output_path
+
+            # Replace previously propagated build outputs with our
+            # main output
+            if [ -e $output_path/nix-support/propagated-build-inputs ]; then
+              rm $output_path/nix-support/propagated-build-inputs
```

Comment:
This sounds quite destructive, the output might have been propagating something else than just `oldOut`. On the other hand, we probably want environments created by `symlinkJoin` to be self-contained so maybe it is okay that propagation is cut-off.

Reply:
Yes, it could be destructive, but for secondary outputs, I could only find `propagated-build-inputs` being added by the multiple outputs hook, not to say that it doesn't exist, but it doesn't seem very prevalent. Propagation shouldn't be cut off for the main output, at least. :)

Reply:
`propagatedBuildInputs` attribute [generates](https://github.com/NixOS/nixpkgs/blob/066fb63b0285cae5dc724ef2fd0b1f6c177e02cf/pkgs/stdenv/generic/setup.sh#L1155) the file in designated `dev` output (in the [order of preference](https://github.com/NixOS/nixpkgs/blob/066fb63b0285cae5dc724ef2fd0b1f6c177e02cf/pkgs/build-support/setup-hooks/multiple-outputs.sh#L32) `$outputDev`, `dev`, `out`).

Reply:
Ah, I see. I had missed that. Then it might be interesting to try copying and mangling the file to replace specifically the main output, but leave the others as-is.

Reply:
It now concatenates all the `paths`' `propagated-build-input` files per output, replacing the previous main output with `$out`. This should be much less destructive.